src/extract.c

```diff
@@ -5,48 +5,124 @@
 #include "backports.h"
 #include <string.h>
 
-int find_offset(SEXP x, SEXP index, int i) {
+int find_offset(SEXP x, SEXP index, int i, int strict) {
   if (Rf_length(index) > 1) {
-    Rf_errorcall(R_NilValue, "Index %i must have length 1", i + 1);
+    Rf_errorcall(R_NilValue,
+      "Index %i must have length 1, not %i.",
+      i + 1,
+      Rf_length(index)
+    );
+  } else if (strict && Rf_length(index) == 0) {
+    Rf_errorcall(R_NilValue, "Index %i must have length 1, not 0", i + 1);
   }
 
   int n = Rf_length(x);
-  if (n == 0)
-    return -1;
+  if (n == 0) {
+    if (strict)
+      Rf_errorcall(R_NilValue, "Object being plucked must have at least one element.");
+    else
+      return -1;
+  }
 
   if (TYPEOF(index) == INTSXP) {
     int val = INTEGER(index)[0];
 
-    if (val == NA_INTEGER)
-      return -1;
+    if (val == NA_INTEGER) {
+      if (strict)
+        Rf_errorcall(R_NilValue,
+          "Index %i must be finite, not NA.",
+          i + 1
+        );
+      else
+        return -1;
+    }
 
     val--;
-    if (val < 0 || val >= n)
-      return -1;
+    if (val < 0) {
+      if (strict)
```

Comment:
General comment: can you add curly braces even to single expressions please?

Reply:
Did you want me to add those for the existing code that I haven't yet modified too?

Reply:
@lionel- I don't think we use that style anymore, and no parens is fine for simple expressions.

Reply:
Ah I wasn' aware, Jim, Gabor and I still use it. It's consistent with our R style and easier to read since we use 2-level indents. Especially with multi-line single expressions such as here. But nevermind then.

Reply:
It can also have implications for covr, I think? That is, the presence of the brackets helps covr do its job in some weird edge cases.

Reply:
I thought the edge cases were fixed and this is officially blessed @jimhester style

Reply:
Maybe I am behind the times then.

Reply:
The officially blessed jimhester style is to always use braces for single expressions.

The edge cases in covr are fixed, but I think it is still better to use braces, it is too easy to accidentally forget to add the braces in the future and write something like the following, then not realize the second if is unconditionally evaluated.
```r
if (val < 0)
  print(val)
  if (strict)
```